wsmaster/che-core-api-factory-github-common/src/main/java/org/eclipse/che/api/factory/server/github/AbstractGithubURLParser.java

```diff
@@ -100,17 +154,31 @@ public GithubUrl parse(String url) throws ApiException {
     return parse(trimEnd(url, '/'), true);
   }
 
+  private IllegalArgumentException buildIllegalArgumentException(String url) {
+    return new IllegalArgumentException(
+        format("The given url %s is not a valid github URL. ", url));
+  }
+
   private GithubUrl parse(String url, boolean authenticationRequired) throws ApiException {
+    Matcher matcher;
     boolean isHTTPSUrl = githubPattern.matcher(url).matches();
-    Matcher matcher = isHTTPSUrl ? githubPattern.matcher(url) : githubSSHPattern.matcher(url);
+    if (isHTTPSUrl) {
+      matcher = githubPattern.matcher(url);
+    } else if (githubSSHPattern.matcher(url).matches()) {
+      matcher = githubSSHPattern.matcher(url);
+    } else {
+      matcher = getPatternMatcherByUrl(url).orElseThrow(() -> buildIllegalArgumentException(url));
+      isHTTPSUrl = url.startsWith("http");
+    }
     if (!matcher.matches()) {
-      throw new IllegalArgumentException(
-          format("The given url %s is not a valid github URL. ", url));
+      throw buildIllegalArgumentException(url);
     }
 
     String serverUrl =
         isNullOrEmpty(oauthEndpoint) || trimEnd(oauthEndpoint, '/').equals(GITHUB_SAAS_ENDPOINT)
-            ? null
+            ? url.startsWith(GITHUB_SAAS_ENDPOINT) || url.startsWith("git@github.com:")
+                ? null
+                : getServerUrl(url).orElseThrow(() -> buildIllegalArgumentException(url))
```

Comment:
The logic here seems overly complicated -- will we ever hit `url.startsWith("git@github.com:")`? 

Maybe we should break this into two separate checks (e.g. if empty, then X, else if matches then etc)

Reply:
Reworked.